### DISCUSSION TITLE
Add Feature to Test + Debug Row Level Security Policies

### DISCUSSION THREAD

**Discussion by Lior539** [supabase/supabase#12269] [Feature Requests]
Currently when writing a RLS policy you have to manually test it. They're also very difficult to debug when things go wrong as there is no logging (the current postgres logs are very cryptic and not suitable for debugging RLS policies.)

Some suggestions I have around this (not sure if they're doable though?):

1. Add the ability for unit tests or test cases for RLS policies
2. Add (better?) logging for debugging RLS policies. e.g. when an API call is made, it would be great to view with what values the RLS policy was executed and what was the output

----
The context behind this request is that Im currently writing a RLS policy that is not behaving as expected and I have no idea how to debug it effectively. I'm just shooting in the dark and modifying my code and rerunning my api calls in the hopes that it works

**Comment by kiwicopple** [comment]
great suggestions @Lior539 - we definitely need to improve our RLS testing (probably with "impersonation" too, so that you can test for a specific user)

**Reply by ozum** [reply]
@kiwicopple. I shared an SQL procedure for impersonation in #12569. I'll be happy if you have a look and advise if it needs improvement or correction.

If you find it useful, please consider including it in Supabase `auth` schema.

Thanks.

**Reply by MildTomato** [reply]
[User impersonation](https://supabase.com/blog/studio-introducing-assistant#user-impersonation) in the dashboard was released last week during Launch Week X

**Reply by vickkhera** [reply]
The User Impersonation feature only helps you interacting with the DB using the REST interface. It doesn't help you when directly connected to the DB and debugging queries, or with pgTAP. The solution [below](#discussioncomment-4908790) by @kjmph works splendidly for these use cases.

**Comment by kylerummens** [comment]
> great suggestions @Lior539 - we definitely need to improve our RLS testing (probably with "impersonation" too, so that you can test for a specific user)

Any timeline for this?

**Reply by MildTomato** [reply]
[User impersonation](https://supabase.com/blog/studio-introducing-assistant#user-impersonation) in the dashboard was released last week during Launch Week X

**Comment by radulescuandrew** [comment]
👀

**Comment by josephbuchma** [comment]
Is it possible to test RLS with pgTap ?
I'm using this helper to set session in my pgTap test:
```sql
CREATE OR REPLACE FUNCTION test_login (user_email text, logout_first boolean = TRUE)
    RETURNS auth.users
    LANGUAGE plpgsql
    AS $$
DECLARE
    auth_user auth.users;
BEGIN
    IF logout_first THEN
        PERFORM
            test_logout ();
    END IF;
    SELECT
        * INTO auth_user
    FROM
        auth.users
    WHERE
        email = user_email;
    PERFORM
        set_config('request.jwt.claim.sub', (auth_user).id::text, TRUE);
    PERFORM
        set_config('request.jwt.claim.role', (auth_user).ROLE, TRUE);
    PERFORM
        set_config('request.jwt.claim.email', (auth_user).email, TRUE);

    PERFORM
        set_config('request.jwt.claims', json_build_object('app_metadata', (auth_user).raw_app_meta_data)::text, TRUE);

    RAISE NOTICE '%', format( 'SET ROLE %I; -- Logging in as %L (%L)', (auth_user).ROLE, (auth_user).id, (auth_user).email);
    EXECUTE format('SET ROLE %I', (auth_user).ROLE);
    RETURN auth_user;
END;
$$;
```

But after logging in as a specific user using `test_login` function, I can still select from a table that should not be accessible to a given user.

**Comment by kjmph** [comment]
Don't you want that to be per session?

```sql
CREATE OR REPLACE FUNCTION test_login (user_email text, logout_first boolean = TRUE)
    RETURNS auth.users
    LANGUAGE plpgsql
    AS $$
DECLARE
    auth_user auth.users;
BEGIN
    IF logout_first THEN
        PERFORM
            test_logout ();
    END IF;
    SELECT
        * INTO auth_user
    FROM
        auth.users
    WHERE
        email = user_email;
    EXECUTE format('SET request.jwt.claim.sub=%I', (auth_user).id::text);
    EXECUTE format('SET request.jwt.claim.role=%I', (auth_user).ROLE);
    EXECUTE format('SET request.jwt.claim.email=%I', (auth_user).email);
    EXECUTE format('SET request.jwt.claims=%I', json_build_object('app_metadata', (auth_user).raw_app_meta_data)::text);

    RAISE NOTICE '%', format( 'SET ROLE %I; -- Logging in as %L (%L)', (auth_user).ROLE, (auth_user).id, (auth_user).email);
    EXECUTE format('SET ROLE %I', (auth_user).ROLE);
    RETURN auth_user;
END;
$$;
```

That seems to function upon my test. Note, I had to `grant authenticated to postgres;`

**Comment by kjmph** [comment]
I also had need of this for testing RLS, and shortened it to the following, and tucked it in auth schema. Not sure if helpful to everyone.

```sql
grant anon, authenticated to postgres;

CREATE OR REPLACE PROCEDURE auth.login_as_user (user_email text)
    LANGUAGE plpgsql
    AS $$
DECLARE
    auth_user auth.users;
BEGIN
    SELECT
        * INTO auth_user
    FROM
        auth.users
    WHERE
        email = user_email;
    EXECUTE format('SET request.jwt.claim.sub=%I', (auth_user).id::text);
    EXECUTE format('SET request.jwt.claim.role=%I', (auth_user).ROLE);
    EXECUTE format('SET request.jwt.claim.email=%I', (auth_user).email);
    EXECUTE format('SET request.jwt.claims=%I', json_strip_nulls(json_build_object('app_metadata', (auth_user).raw_app_meta_data))::text);

    RAISE NOTICE '%', format( 'SET ROLE %I; -- Logging in as %L (%L)', (auth_user).ROLE, (auth_user).id, (auth_user).email);
    EXECUTE format('SET ROLE %I', (auth_user).ROLE);
END;
$$;

CREATE OR REPLACE PROCEDURE auth.login_as_anon ()
    LANGUAGE plpgsql
    AS $$
BEGIN
    EXECUTE format('SET request.jwt.claim.sub=''''');
    EXECUTE format('SET request.jwt.claim.role=''''');
    EXECUTE format('SET request.jwt.claim.email=''''');
    EXECUTE format('SET request.jwt.claims=''''');
    EXECUTE format('SET ROLE anon');
END;
$$;

CREATE OR REPLACE PROCEDURE auth.logout ()
    LANGUAGE plpgsql
    AS $$
BEGIN
    EXECUTE format('SET request.jwt.claim.sub=''''');
    EXECUTE format('SET request.jwt.claim.role=''''');
    EXECUTE format('SET request.jwt.claim.email=''''');
    EXECUTE format('SET request.jwt.claims=''''');
    EXECUTE format('SET ROLE postgres'); -- For our use cases, this is 'postgres'. This may not be the case for everyone.
END;
$$;
```

**Reply by chaoky** [reply]
can this be safely added to a production database?

**Reply by christophemarois** [reply]
Extremely useful, thank you.

**Reply by imageck** [reply]
@kjmph does not seem to work in the SQL editor? After calling either of the procedures, my `current_role` is still set to `postgres` and the `request.jwt` settings return null.

**Reply by GaryAustin1** [reply]
https://github.com/GaryAustin1/RLS-Performance#tools-to-measure-performance  Towards the end of this shows a shorter way to test in the SQL editor.

**Comment by tvogel** [comment]
> ```sql
>     EXECUTE format('SET request.jwt.claim.sub=%I', (auth_user).id::text);
>     EXECUTE format('SET request.jwt.claim.role=%I', (auth_user).ROLE);
>     EXECUTE format('SET request.jwt.claim.email=%I', (auth_user).email);
>     EXECUTE format('SET request.jwt.claims=%I', json_strip_nulls(json_build_object('app_metadata', (auth_user).raw_app_meta_data))::text);
> ```

Because I have lengthy `app_metadata`, I changed some of the `%I` to `%L` because `psql` complained, it had to truncate the "identifier". I am a Postgres newbie still. Will this cause problems?

```sql
    EXECUTE format('SET request.jwt.claim.sub=%L', (auth_user).id::text);
    EXECUTE format('SET request.jwt.claim.role=%I', (auth_user).ROLE);
    EXECUTE format('SET request.jwt.claim.email=%L', (auth_user).email);
    EXECUTE format('SET request.jwt.claims=%L', json_strip_nulls(json_build_object('app_metadata', (auth_user).raw_app_meta_data))::text);
```

**Comment by kjmph** [comment]
Ah, good catch. That's why I threw in the `json_strip_nulls`; on second thought we should probably use `''%s''`; I changed my local version.

**Reply by vickkhera** [reply]
That won't protect you from embedded single quotes in the data (say hello to [Bobby Tables](https://xkcd.com/327/)). The `%L` format is the right one.

**Comment by TomasHubelbauer** [comment]
Is there a way to at least tell which policy was violated? I am looking at an error that says "new row violates row-level security policy for table" and I have nothing to go on in terms of debugging other than bisecting the conditions of the policy, but that takes a lot of time because it has to do with a specific user whose auth I can only get by using the web app to call the Supabase API where the policy is not behaving as expected. I can't work faster by using Supabase Studio or Postico because the auth is different there.

**Reply by Evaldas-B** [reply]
Your Row Level Security (RLS) rules probably rely on data stored in the JWT (JSON Web Token) to verify if a user has the right permissions for any request they make. The JWT is sent with every request to the server. To easily check who is making a request and what access they should have, you can create a Remote Procedure Call (RPC) to retrieve this information.

1. **Create the RPC:**
   ```sql
   CREATE OR REPLACE FUNCTION who_am_i() RETURNS TEXT AS $$
     SELECT auth.jwt();
   $$ LANGUAGE SQL;
   ```

2. **Check the data attached to the JWT:**
   ```typescript
   const { data: whoAmI} = await supabase.rpc("who_am_i");
   console.log(JSON.stringify(whoAmI, null, 2));
   ```

**Comment by aaronksaunders** [comment]
other than for pretty basic policies, I don't know how you can really be productive with RLS without some UI to test the policies and some way to impersonate a user. Been trying to get RLS working today and just gave up and wrote way too much javascript code to solve my problem.

**Reply by jkazimierczak** [reply]
I agree with the impersonation part - keeping in mind what policy does what is just tiring. A solution where maybe we could just set userID with some variable to impersonate would be useful in testing the policies. I don't have any opinion about the UI, as queries can be executed just with SQL.

**Reply by diemer** [reply]
I found this method to be really straightforward for the app I'm building: https://dev.to/akkilah/how-to-implement-role-based-access-with-supabase-3a2

You can almost impersonate using the `get_claims` function

**Comment by munaf-khatri** [comment]
RLS is the security boundary, but hand-writing pgTAP tests for it is tedious enough that most policies never get tested. Supabase's own docs even call it "inaccessible to most web developers."
I built a small open-source tool for exactly this: rlsautotest (Apache-2.0). You point it at a disposable copy of your DB and it:

reads your RLS policies from the catalog
generates the pgTAP tests and the seed data, per table, per command, per identity (owner, other user, other tenant, anon, role-holder)
proves who can SELECT / INSERT / UPDATE / DELETE which rows
emits a --report / --html access matrix, plus a CI gate that fails the build on a leak or an unprotected table

What took the most work was making a green test actually mean something:

it seeds a row that's really owned by the identity under test (works backward from each policy), so passing proves real isolation, not an empty-table pass
anything it can't verify soundly, it marks instead of faking

Repo: https://github.com/unitautogen/rlsautotest
Would really value feedback from folks here, and happy to run it against a policy set if anyone wants to see what it catches.
